In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.utils import resample

pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings('ignore')

# Random Forest

In [2]:
numerical = pd.read_csv('numerical.csv')
categorical = pd.read_csv('categorical.csv')
targets = pd.read_csv('target.csv')
data = pd.concat([numerical, categorical, targets], axis = 1)

In [3]:
data.head()

TCODE        AGE  INCOME  WEALTH1  HIT  MALEMILI  MALEVET  VIETVETS  \
0      0  60.000000       5        9    0         0       39        34   
1      1  46.000000       6        9   16         0       15        55   
2      1  61.611649       3        1    2         0       20        29   
3      0  70.000000       1        4    2         0       23        14   
4      0  78.000000       3        2   60         1       28         9   

   WWIIVETS  LOCALGOV  STATEGOV  FEDGOV  WEALTH2  POP901  POP902  POP903  \
0        18        10         2       1        5     992     264     332   
1        11         6         2       1        9    3611     940     998   
2        33         6         8       1        1    7001    2040    2669   
3        31         3         0       3        0     640     160     219   
4        53        26         3       2        9    2520     627     761   

   POP90C1  POP90C2  POP90C3  POP90C4  POP90C5  ETH1  ETH2  ETH3  ETH4  ETH5  \
0        0       35       65       47       53    92     1     0     0    11   
1       99        0        0       50       50    67     0     0    31     6   
2        0        2       98       49       51    96     2     0     0     2   
3        0        8       92       54       46    61     0     0    11    32   
4       99        0        0       46       54     2    98     0     0     1   

   ETH6  ETH7  ETH8  ETH9  ETH10  ETH11  ETH12  ETH13  ETH14  ETH15  ETH16  \
0     0     0     0     0      0      0      0     11      0      0      0   
1     4     2     6     4     14      0      0      2      0      1      4   
2     0     0     0     0      0      0      0      2      0      0      0   
3     6     2     0     0      0      0      0     31      0      0      1   
4     0     0     0     0      0      0      0      0      0      0      0   

   AGE901  AGE902  AGE903  AGE904  AGE905  AGE906  AGE907  CHIL1  CHIL2  \
0      39      48      51      40      50      54      25     31     42   
1      34      41      43      32      42      45      32     33     46   
2      35      43      46      37      45      49      23     35     40   
3      32      40      44      34      43      47      25     45     35   
4      33      45      50      36      46      50      27     34     43   

   CHIL3  AGEC1  AGEC2  AGEC3  AGEC4  AGEC5  AGEC6  AGEC7  CHILC1  CHILC2  \
0     27     11     14     18     17     13     11     15      12      11   
1     21     13     14     33     23     10      4      2      11      16   
2     25     13     20     19     16     13     10      8      15      14   
3     20     15     25     17     17     12      7      7      20      17   
4     23     14     21     13     15     20     12      5      13      15   

   CHILC3  CHILC4  CHILC5  HHAGE1  HHAGE2  HHAGE3  HHN1  HHN2  HHN3  HHN4  \
0      34      25      18      26      10      23    18    33    49    28   
1      36      22      15      12       1       5     4    21    75    55   
2      30      22      19      25      10      23    21    35    44    22   
3      30      14      19      25      11      23    23    27    50    30   
4      34      19      19      31       7      27    16    26    57    36   

   HHN5  HHN6  MARR1  MARR2  MARR3  MARR4  HHP1  HHP2  DW1  DW2  DW3  DW4  \
0    12     4     61      7     12     19   198   276   97   95    2    2   
1    23     9     69      4      3     24   317   360   99   99    0    0   
2     6     2     63      9      9     19   183   254   69   69    1    6   
3    15     8     63      9      6     23   199   283   85   83    3    4   
4    24    14     42     17      9     33   235   323   99   98    0    0   

   DW5  DW6  DW7  DW8  DW9   HV1   HV2  HV3  HV4  HU1  HU2  HU3  HU4  HU5  \
0    0    0    7    7    0   479   635    3    2   86   14   96    4    7   
1    0    0    0    0    0  5468  5218   12   10   96    4   97    3    9   
2    5    3    3    3    0   497   546    2    1   78   22   93    7   18   
3  

In [4]:
list(data.isna().sum())

[0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,


I can see that I have two nulls in the third column starting by the end so I will eliminate those 2 rows

In [5]:
data[data['FIRSTDATE_MM'].isna()]

TCODE        AGE  INCOME  WEALTH1  HIT  MALEMILI  MALEVET  VIETVETS  \
83217      2  61.611649       1        7    3         0       31        34   
93811      0  78.000000       5        5    2         0       29        34   

       WWIIVETS  LOCALGOV  STATEGOV  FEDGOV  WEALTH2  POP901  POP902  POP903  \
83217        30         5         2       2        9    1250     368     446   
93811        52         1         0       7        5     935     237     341   

       POP90C1  POP90C2  POP90C3  POP90C4  POP90C5  ETH1  ETH2  ETH3  ETH4  \
83217        0        0       99       50       50    99     0     0     0   
93811       99        0        0       44       56    98     1     0     1   

       ETH5  ETH6  ETH7  ETH8  ETH9  ETH10  ETH11  ETH12  ETH13  ETH14  ETH15  \
83217     0     0     0     0     0      0      0      0      0      0      0   
93811     1     0     0     0     0      0      0      0      0      0      0   

       ETH16  AGE901  AGE902  AGE903  AGE904  AGE905  AGE906  AGE907  CHIL1  \
83217      0      34      43      46      36      46      49      28     33   
93811      0      38      47      51      42      51      54      21     38   

       CHIL2  CHIL3  AGEC1  AGEC2  AGEC3  AGEC4  AGEC5  AGEC6  AGEC7  CHILC1  \
83217     42     25     11     19     21     17     13     11      8      11   
93811     42     20      9     20     18     11     12     11     20      17   

       CHILC2  CHILC3  CHILC4  CHILC5  HHAGE1  HHAGE2  HHAGE3  HHN1  HHN2  \
83217      16      32      25      17      26       9      24    15    37   
93811      14      33      21      15      25      12      24    27    34   

       HHN3  HHN4  HHN5  HHN6  MARR1  MARR2  MARR3  MARR4  HHP1  HHP2  DW1  \
83217    48    30    12     4     70      6      5     19   194   279   92   
93811    39    22     6     2     53     12     17     18   168   242   92   

       DW2  DW3  DW4  DW5  DW6  DW7  DW8  DW9  HV1  HV2  HV3  HV4  HU1  HU2  \
83217   91    2    3    2    1    0    0    0  544  589    2    1   83   17   
93811   92    3    7    4    3   12   12    0  389  399    3    3   83   17   

       HU3  HU4  HU5  HHD1  HHD2  HHD3  HHD4  HHD5  HHD6  HHD7  HHD8  HHD9  \
83217   96    4    5    40    83    75    36    91     9     4     1     3   
93811   97    3    9    35    70    55    25    75    25     9     1     9   

       HHD10  HHD11  HHD12  ETHC1  ETHC2  ETHC3  ETHC4  ETHC5  ETHC6  HVP1  \
83217     10     11      2     23     58     18      0      0      0     0   
93811     12     23      4     18     52     28      0      0      0     0   

       HVP2  HVP3  HVP4  HVP5  HVP6  HUR1  HUR2  RHP1  RHP2  RHP3  RHP4  \
83217     1     7    23    57     0     1    62    60    62    15     4   
93811     0     0     1    17     0     0    38    49    53    13     4   

       HUPA1  HUPA2  HUPA3  HUPA4  HUPA5  HUPA6  HUPA7  RP1  RP2  RP3  RP4  \
83217      3      0      3     13      2      1      1    2    6   22   61   
93811      4      3      0     11      2      3      0    0   20   77   91   

          MSA    ADI    DMA  IC1  IC2  IC3  IC4    IC5  IC6  IC7  IC8  IC9  \
83217  7720.0  391.0  624.0  314  390  371  421  14581   21   21   11   23   
93811  3720.0   59.0  563.0  349  379  371  411  12908   14    7   30   28   

       IC10  IC11  IC12  IC13  IC14  IC15  IC16  IC17  IC18  IC19  IC20  IC21  \
83217    18     2     0     0     4    12    20    13    26    22     2     0   
93811    17     5     0     0     0     7     9    25    29    23     6     0   

       IC22  IC23  HHAS1  HHAS2  HHAS3  HHAS4  MC1  MC2  MC3  TPE1  TPE2  \
83217     0     5     35      3     59      7   28   72    8    71    13   
93811     0     0     34      4     46      1   47   53   10    96     2   

       TPE3  TPE4  TPE5  TPE6  TPE7  TPE8  TPE9  PEC1  PEC2  TPE10  TPE11  \
83217     0     0     0     0     0     6     9     5     2     19     20   
93811     0     0     0     0     0     0     2     0    12     13     16  

In [6]:
data = data.drop([83217,93811], axis=0).reset_index()

In [7]:
data['TARGET_B'].value_counts()

0    90567
1     4843
Name: TARGET_B, dtype: int64

As the target is unbalanced, I will do upsampling

In [8]:
y = data['TARGET_B']
X = data.drop(['TARGET_B'], axis = 1)

numericalX = X.select_dtypes(np.number)
categoricalX = X.select_dtypes(np.object)

#I'll use OneHotEncode in the categoricals
from sklearn.preprocessing import OneHotEncoder
encoder = OneHotEncoder(drop='first').fit(categoricalX)
encoded_categorical = encoder.transform(categoricalX).toarray()
encoded_categorical = pd.DataFrame(encoded_categorical)
X = pd.concat([numericalX, encoded_categorical], axis = 1)

#Doing train/test split before upsampling, and then only upsample the training set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)


In [9]:
#concatenating X_train and y_train to upsample them
trainset = pd.concat([X_train, y_train], axis=1)

# upsampling category 1:
category_0 = trainset[trainset['TARGET_B'] == 0]
category_1_normal = trainset[trainset['TARGET_B'] == 1]
category_1 = resample(category_1_normal, 
                                   replace=True,
                                   n_samples = len(category_0))
print(category_1.shape)

category_0 = trainset[trainset['TARGET_B']== 0 ]
trainset_new = pd.concat([category_0, category_1], axis = 0)
trainset_new = trainset_new.sample(frac =1) #randomize the rows
X_train = trainset_new.drop(['TARGET_B'], axis=1)
y_train = trainset_new['TARGET_B']
print(X_train.shape)

(72496, 357)
(144992, 356)


The rows from X_train are the double of category_1 upsampled

In [10]:
X_train = pd.DataFrame(X_train)
X_test = pd.DataFrame(X_test)

#Saving the column target_D for next steps
y_train_regression = X_train['TARGET_D']
y_test_regression = X_test['TARGET_D']

# Removing the column target D from the set of features
X_train = X_train.drop(['TARGET_D'], axis = 1)
X_test = X_test.drop(['TARGET_D'], axis = 1)

In [11]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix

clf = RandomForestClassifier(max_depth=5,
                             min_samples_split=20,
                             min_samples_leaf =20,
                             max_samples=0.2,
                             random_state = 42)
clf.fit(X_train, y_train)
print("Train score: ", clf.score(X_train, y_train))
print("Test score: ", clf.score(X_test, y_test))

y_pred = clf.predict(X_test)
display(y_test.value_counts())
display(confusion_matrix(y_test, y_pred))

0.6225791767821673
0.6072214652552144


0    18071
1     1011
Name: TARGET_B, dtype: int64

array([[11012,  7059],
       [  436,   575]])

#### Calculating costs

In [12]:
costs = {"TARGET_D": y_test_regression, 
        "TARGET_B": y_test}

In [13]:
costs_df = pd.concat(costs, axis = 1)

In [14]:
costs_df['TARGET_D'].describe()

count    19082.000000
mean         0.837361
std          4.637765
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max        200.000000
Name: TARGET_D, dtype: float64

In [15]:
costs_df[costs_df['TARGET_B']==1]['TARGET_D'].describe()

count    1011.000000
mean       15.804679
std        13.021494
min         1.000000
25%        10.000000
50%        13.000000
75%        20.000000
max       200.000000
Name: TARGET_D, dtype: float64

In [16]:
19082 * (0.837 - 0.68) # number of mails * (mean donation for all - cost of mailing)

2995.8739999999984

If the company sends a mail to all the clients, the profit will be: $2996

In [18]:
(575 * 15.804) - ((575 + 436) * 0.68) # number of mails * mean donation for those who have donated - (cost of mailing from those who have donated and not)

8399.82

If the company sends a mail to those clients that the algorith has returned as donnors , the profit will be: $8400

The difference is quite important